src/main/java/com/google/cloud/tools/appengine/cloudsdk/process/LegacyProcessHandler.java

```diff
@@ -0,0 +1,274 @@
+              }
+            }
+            stdOut.close();
+          }
+        };
+    stdOutThread.setDaemon(true);
+    stdOutThread.start();
+    return stdOutThread;
+  }
+
+  private Thread handleErrOut(Process process) {
+    final Scanner stdErr = new Scanner(process.getErrorStream(), Charsets.UTF_8.name());
+    Thread stdErrThread =
+        new Thread("standard-err") {
+          @Override
+          public void run() {
+            while (stdErr.hasNextLine() && !Thread.interrupted()) {
+              String line = stdErr.nextLine();
+              for (ProcessOutputLineListener stdErrLineListener : stdErrLineListeners) {
+                stdErrLineListener.onOutputLine(line);
+              }
+            }
+            stdErr.close();
+          }
+        };
+    stdErrThread.setDaemon(true);
+    stdErrThread.start();
+    return stdErrThread;
+  }
+
+  private void syncRun(Process process, Thread stdOutThread, Thread stdErrThread)
+      throws InterruptedException, AppEngineException {
+    int exitCode = process.waitFor();
+    // https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/269
+    if (stdOutThread != null) {
+      stdOutThread.join();
+    }
+    if (stdErrThread != null) {
+      stdErrThread.join();
+    }
+
+    for (ProcessExitListener exitListener : exitListeners) {
+      exitListener.onExit(exitCode);
+    }
+  }
+
+  private void asyncRun(
+      final Process process, final Thread stdOutHandler, final Thread stdErrHandler)
+      throws ProcessHandlerException {
+    if (!exitListeners.isEmpty()
+        || !stdOutLineListeners.isEmpty()
+        || !stdErrLineListeners.isEmpty()) {
+      Thread exitThread =
+          new Thread("wait-for-process-exit-and-output-handlers") {
+            @Override
+            public void run() {
+              try {
+                syncRun(process, stdOutHandler, stdErrHandler);
+              } catch (InterruptedException | AppEngineException e) {
+                e.printStackTrace();
+              }
+            }
+          };
+      exitThread.setDaemon(true);
+      exitThread.start();
+      try {
+        waitingProcessOutputLineListener.await();
+      } catch (ProcessRunnerException e) {
+        // TODO: remove this when we remove process runner exception and process runner (in
+        // TODO: followup PR)
+        throw new ProcessHandlerException(e);
+      }
+    }
+  }
+
+  private static void shutdownProcessHook(final Process process) {
+    Runtime.getRuntime()
+        .addShutdownHook(
+            new Thread("destroy-process") {
+              @Override
+              public void run() {
+                if (process != null) {
+                  process.destroy();
+                }
+              }
+            });
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private List<ProcessOutputLineListener> stdOutLineListeners = new ArrayList<>();
```

Comment:
These defaults could probably be removed so that the Lists can be final, or just have `new DevAppServerAsyncOutputWatcherFactory()` be a default as well.